### DISCUSSION TITLE
Trace id changes while creating custom span

### DISCUSSION THREAD

**Discussion by balkrishan333** [open-telemetry/opentelemetry-java#3905] [Q&A] [answered]
For adding business information to my spans I am creating a manual span (this is because I think I can't add attributes to span obtained by Span.current()) and adding required information to that but I noticed that my default span and span which I created manually have different trace ids. I was expecting trace id for both to be same. I am using following code to create manual span

```
        OpenTelemetry openTelemetry = OpenTelemetryFactory.getInstance();
        Tracer tracer = openTelemetry.getTracer("");

        Span parentSpan = Span.current();
        Span span = tracer.spanBuilder("balkrishan span").setParent(Context.current().with(parentSpan))
                .setSpanKind(SpanKind.SERVER).startSpan();

        span.setAttribute("messageType", "v2");
        span.setAttribute("tenant", "dummy tenant");
       
        // put the span into the current Context
        try (Scope scope = span.makeCurrent()) {
            // app code
             
        } catch (Throwable t) {
            span.setStatus(StatusCode.ERROR, "Change it to your error message");
        } finally {
            span.end(); // closing the scope does not end the span, this has to be done manually
        }
``` 

I am using jaeger as visualisation tool. I can see both the spans in jaeger but trace id for both of them is different.
Please help me identify what I am doing wrong or is my expectation that trace id should be same unfair.

Just to be clear, when I say default span, my method's entry point is "/tracing" and I can see 1 span with name "/tracing" and another one named "balkrishan span" as created by code above.

**Comment by jkwatson** [comment, marked as answer]
1. You definitely can add attributes to an existing Span. 
2. Since there is already a Span in the current Context, you don't need to put it in again when parenting your new span. Just don't manually specify a parent on your new Span and it'll use the one already in the current Context.

**Reply by balkrishan333** [reply]
Thanks for reply @jkwatson . I tried that already but new attributes are not reflected when I print span.

Here is the code which I used:

```
        Span span2 = Span.current();
        span2.setAttribute("messageType", "v2");
        span2.setAttribute("tenant", "dummy tenant");
        System.out.println("span = " + span2);
```
When I print the span there are no additional attributes present in span. Below is the result of printing the span:

> span = PropagatedSpan{ImmutableSpanContext{traceId=00000000000000000000000000000000, spanId=0000000000000000, traceFlags=00, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=false}}

When I saw this, I had 2 observations:

1. valid is set to false, does this mean span is not valid. could this be a reason for new attributes not being reflecting.
2. Span is of type PropagatedSpan, when I looked the source of setAttribute of PropagatedSpan it does nothing. Below is the code snippet from PropagatedSpan

```
  @Override
  public Span setAttribute(String key, String value) {
    return this;
  } 
```

Can you please suggest what am I doing wrong over here.

**Reply by jkwatson** [reply]
Ah... if the span that's in the context is a propagated one, then yes, you can't modify that one. (couldn't tell that from your code snippet).

**Reply by balkrishan333** [reply]
Thanks for confirming. Coming back to my original question, can you please help why my new span gets a new trace is rather than using existing one.

**Reply by jkwatson** [reply]
Did you see my second comment: 

> Since there is already a Span in the current Context, you don't need to put it in again when parenting your new span. Just don't manually specify a parent on your new Span and it'll use the one already in the current Context.

**Reply by jkwatson** [reply]
Is your second span being created on another thread? Can you provide a reproduction case that I can look at?

**Reply by balkrishan333** [reply]
> Did you see my second comment:
> 
> > Since there is already a Span in the current Context, you don't need to put it in again when parenting your new span. Just don't manually specify a parent on your new Span and it'll use the one already in the current Context.

Yes, tried that as well but nothing changes. Both span still have different trace id. The method in which I am creating the span is entry point for my request in rest controller, does that make any difference? I assume it should not make any difference but just thought of checking about it.

**Reply by jkwatson** [reply]
Do you have a reproduction case we can look at?

**Reply by balkrishan333** [reply]
I have push the code to following repo:

https://github.com/balkrishan333/playground-app

Use build.sh file to use docker compose to run app and jaeger. 

Access jaeger on http://localhost:16686
For generating traces, use, http://localhost:8080/tracing (If you hit this URL twice there will be error, this is due to how I am using opentelemetry object but hitting the url once is enough to replicate the issue)

On jaeger UI, there are 2 spans, one named "/tracing" and other named "customSpan". I am expecting the same trace id for these 2 spans.

For code, refer method named tracing in SampleController.java class. I am creating customSpan in this method.

To see how I am creating OpenTelemetry object refer class, OpenTelemetryFactory.java

Let me know if some thing is not clear or if you want me add more information.

**Reply by jkwatson** [reply]
I took a look at your sample app, and there are quite a few issues to be sorted out.

1) If you're using the javaagent, you cannot configure your own SDK; you must use the GlobalOpenTelemetry instance that the agent provides you.
2) You *must* align your versions. It looks like you have version 0.14 of the javaagent, which is absolutely in no way compatible with recent versions of the APIs or SDKs.  Also, you have a mix of versions of the various libraries that you're using.

**Reply by balkrishan333** [reply]
Thanks for the reply. I tried both the suggestions and here is the out come

1. I updated the javaagent version to 1.7.2 and I also tried to use correct versions of other libs. For all libs other than opentelemetry-exporter-otlp I am using version 1.7.0 and for opentelemetry-exporter-otlp I am using 1.8.0-SNAPSHOT (This is because version 1.7.0 for opentelemetry-exporter-otlp available on web is not correct and does not have any class files, so I have built the jar locally from source code from 1.7.0 branch. Though I have built it form 1.7.0 branch but not sure why version is 1.8.0-SNAPSHOT)

After making the above change, I can see that my customSpan has the same trace id as that of current span which was not the case earlier but my default span is not not being exported and I can see below error on console:

> playground-app_1  | [otel.javaagent 2021-11-23 11:27:30:355 +0000] [grpc-default-executor-1] ERROR io.opentelemetry.exporter.otlp.internal.grpc.DefaultGrpcExporter - Failed to export spans. Server is UNAVAILABLE. Make sure your collector is running and reachable from this network. Full error message:UNAVAILABLE: io exception

2. If I use GlobalOpentelemetry object, none of my span is exported and I get no error on console as well.

So, after updating the version of java agent, things seem to be going in correct directions but now the problem is my spans from default spans are not being exported. 

Can you please suggest, do I have to do something different with new version of java agent in order for my spans to be exported.

Also, to make sure default span not being exported is not due to my custom span, I removed all the code for creating cutom span but even then default span is not being exported.

**Reply by balkrishan333** [reply]
Hi

I got it working. After upgrading the javaagent version, I needed to make some changes in parameter names. Once correct parameter names where used, everything is working as per expectation. My original problem of different trace id with 2 spans is also solved. Only thing is GlobalOpenTelemetry does not work for me.

Thanks for all your support.

**Reply by jkwatson** [reply]
Have you updated your example code with the example of `GlobalOpenTelemetry` not being properly initialized by the javaagent?

**Reply by balkrishan333** [reply]
I am not sure what you meant but I have used followed code while trying to use `GlobalOpenTelemetry`

```
        OpenTelemetry globalOpenTelemetry = GlobalOpenTelemetry.get();
        Tracer tracer = globalOpenTelemetry.getTracer("");
```

Do you mean I need to take care of some initiailization in order to use `GlobalOpenTelemetry`

**Reply by jkwatson** [reply]
no, I was looking at the example you provided, and it still had SDK initialization code in it, so I assume you hadn't updated it for me to see what problem you might be encountering.

**Reply by balkrishan333** [reply]
Yeah, you are right. When trying with `GlobalOpenTelemetry` I didn't realize that I am initialising the SDK and just tried to use `GlobalOpenTelemetry` after removing calls to SDK opentelemetry object. Just now I tried with `GlobalOpenTelemetry` after removing the initialization code for SDK and it works without any problem.

Thanks for your help.
